### DISCUSSION TITLE
Help LUT calibration

### DISCUSSION THREAD

**Discussion by FreubelNL** [awawa-dev/HyperHDR#375] [General]
So i have the awesome hyperhdr running with an USB grabber. Works like a charm. Unfortunatly i think i cheaped out on the lightstrip because colors are pretty off. So i decided to order a new one from BTF lighting. Meanwhile i have been reading up on the LUT calibration as it is mentioned in the hyperhdr webpage. And i can't figure it out. Is there any step to follow tutorial for the LUT callibration ?
I have hyperhdr running on a raspberry pi 4 with the USB grabber.

**Comment by awawa-dev** [comment]
Hi
I think the steps are pretty good explained at least from my point of view. Of course it's possible to run into the troubles if you run it for first time. I recommend read this discussion topic to avoid common mistakes (expand the discussion): https://github.com/awawa-dev/HyperHDR/discussions/329 And what is your USB grabber model?

**Reply by FreubelNL** [reply]
Hia awawa-dev, thank you for responding.
Regarding the explaining in the webpage - i have been desiging and selling bathrooms for over 20 years, for me it's a piece of cake. Pretty much everything is easy if you know how to do it. Honestly, i have no idea how to run it. Can it be run from a chromebook for instance ? Do i need a dedicated win10 machine ? etc. etc. I will dive into the discussion and see if i can make anything from it. Thank you for the link !
Regarding the USB grabber, i hope it's oke to drop a link from amazon here, if not feel free to delete it.
[USB grabber](https://www.amazon.nl/gp/product/B08SBTK4QH/ref=ppx_yo_dt_b_asin_title_o02_s00?ie=UTF8&psc=1)
As far as i can judge it it's pretty much the same as the one that the smart home resolver video on youtube is using. Just a different brand.
If it's not good enough for hyperhdr please let me know. I do not use cec hence the USB version.

*edit*
I have read the discussion but it is still not clear for me. I will explain my situation. I have a RPI4 which runs on raspian lite, it's also headless. So layout would be:
Denon receiver 4K hdmi out -> USB grabber -> RPI4 -> LEDstrip while the output from the grabber goes to my tv (LG Oled tv)
So for me it is not clear how to project the image to the tv. I assume i can't use the tv's webbrowser and open the hyperhdr webpage and run it from there because the capturing doesn't work that way. So i have to present it before it goes into the grabber. So can i use the HDMI output from a win10 laptop, open the webpage and run it from there and feed that into the grabber ?

**Comment by awawa-dev** [comment]
Judging by the price, I'm not sure the grabber is capable of forwarding/capturing an HDR signal. Usually grabbers with a loop that are HDR capable are more expensive. Probably cheapest is 4KVC00 (currently $60, jumped $10 recently) but has it's own problems. Alternative and cheaper solution is using MS2109 grabber and an HDMI splitter (also HDCP capable like Ezcoo EZ-SP12H2, it forwards SDR to MS2109 and HDR to the TV).

For the calibration in your case you need two units (because Rpi is headless, and I doubt it can output HDR anyway,  it can be calibrated using one Windows computer running HyperHDR with a grabber connected to itself, but I skip this scenario to keep it simple)

1) Windows PC, the video output is connected to the grabber that is captured by Rpi. TV is alse receiving and displaying the signal.
2) on the Windows PC:  you enable forced HDR output in the graphic properties. Your TV should confirm it receives HDR signal but it depends on the model. 
3) If your Windows tells you in the graphic properties that the display is not HDR capable (TV may be capable but the grabber is limiting its options) then unfortunately the calibration is not possibly for this setup.
4) on the Windows PC: open web browser using HyperHDR address running on the Rpi4.
5) on the Windows PC: make sure in the live preview in HyperHDR that it's capturing the screen from PC Windows
6) on the Windows PC: in the browser open the HyperHDR calibration page, press F11 to make it full-screen and finally click the 'Start' button

**Reply by FreubelNL** [reply]
Thank you very much ! This is a really helpfull tutorial.
I just ordered the following parts:
[USB grabber](https://www.amazon.nl/KDR-Streaming-Onderwijs-Videoconferentie-Broadcasting/dp/B08QCR5C5Z/ref=sr_1_26?__mk_nl_NL=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1S80HIT5O50HI&keywords=MS2109&qid=1665030086&qu=eyJxc2MiOiIwLjAwIiwicXNhIjoiMC4wMCIsInFzcCI6IjAuMDAifQ%3D%3D&sprefix=ms2109%2Caps%2C341&sr=8-26&th=1)
And the [HDMI splitter](https://www.amazon.nl/HDMI-Splitter-1x2-Mini-60Hz/dp/B083JVPXDY/ref=sr_1_1?__mk_nl_NL=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1L6Z9U0GOZKB1&keywords=EZ-SP12H2&qid=1665030264&qu=eyJxc2MiOiIwLjk1IiwicXNhIjoiMC4wMCIsInFzcCI6IjAuMDAifQ%3D%3D&sprefix=ez-sp12h2%2Caps%2C318&sr=8-1)
If you can just check it ? 
And i think you are absolutly spot on on the HDR signal, i can't recall having seen the HDR logo on the tv the last few days even though i was watching some movies to try out the ambi lights which are HDR formats.

**Reply by awawa-dev** [reply]
I don't know this grabber. Looks like a "fake" MS2109 clone, because MS2109 is USB2.0 grabber and using USB3.0 connector can't change this fact. Look on my blog, there is USB2.0 base implementation made by Rullz and some other manufactures. It's already a cheap grabber and never know what you get because the next revisions can bring some changes inside, but the problem with custom designs is sometimes they have problems with thermals (missing internal radiators) and sometimes there are small changes in the firmware, but they can affect the quality (or at least they need a calibration).
The HDMI splitter is OK, many users recommend it.

**Reply by FreubelNL** [reply]
Agreed, problem is though, i can't get that specific grabber here in NL. So i went for the one that appears to be the same.
I'll just have to wait and see if it transfers the HDR signal.

**Reply by awawa-dev** [reply]
I think the grabber you chose will work fine, it's simply safer to choose tested models because some users reported problems with their version of MS2109. On the page you sent, I see a similar model to Rullz version, probably it's the same inside https://www.amazon.nl/KDR-Streaming-Onderwijs-Videoconferentie-Broadcasting/dp/B08QCDYM4G so if you encounter the problems you can replace it.

The grabber won't interfere with the HDR signal for sure, if you connect it to the HDMI splitter port 2 (and for this Ezcoo model enable a scaller-switch for port 2). TV as higher-end device, should be connected to port 1 according to the manual (it's usually the same way for other HDMI splitters with a scaler).  
Thanks to the splitter TV will receive HDR and the grabber won't (but the colors will be washed out, HyperHDR will fix it).

**Reply by FreubelNL** [reply]
Thats pretty much my idea to. Let 's what happens and if it works it works and if it doesn't i can always look for a different one.

**Comment by awawa-dev** [comment]
And in case of MS2109 the calibration may not be necessary. I made the calibration myself and it's shared to download in the HyperHDR release section. You must only follow the short manual inside the archive that will tell you, how to setup brightness/contrast/hue etc  in HyperHDR because it's important for this grabber (the factory settings are not neutral and distort the video).

**Reply by FreubelNL** [reply]
Again, thank you very much, i have downloaded the LUT file and will drop it on the SD card that's in the RPI.
Can't wait for the weekend to switch the stuff around and enjoy this even more.
There isn't a lot a 52 year old dude can look forward to !

*edit*
So in prepartion i have been looking into the LUT part and the pre-fab tables.
I assume i change those recommended settings on this page right ?
![Screenshot 2022-10-06 07 10 59](https://user-images.githubusercontent.com/87243051/194219955-a3ecf3b5-ba8d-4a30-8755-7dc4ebdac1d9.png)

**Reply by awawa-dev** [reply]
Exactly. For MS2109 grabber set hardware brightness to 1, contrast and saturation to 128. Good luck!

**Comment by FreubelNL** [comment]
I received the 2 pieces of hardware today so i'm able to redo the entire thing this sunday !
Looking forward to that.
Maybe a weird question but i 'm also changing the led strip to the sk6812 version, I assume that doesn't make any difference regarding the LUT table ? Or would you recommend the ws2812b strip ?

**Reply by awawa-dev** [reply]
LUT table is still the same for sk6812 RGBW or any other type of LED strip. It may only make a difference if you drive the LED strip from Rpi directly (which is not officially supported) because it's more tricky to maintain stable protocol: one byte per led more to transfer, may not work for comparable number of working ws2812b leds.

**Reply by FreubelNL** [reply]
I do drive them directly from the RPI without any esp or other hardware, i take it you mean those and not a powerbrick, because the led and the rpi both have their own powersupply. How can i tell if i do need to add the esp ?
If i look at the logs right now the only line i pretty much see is this one:
`2022-10-07T13:13:00.053Z [PERFORMANCE] [USB capturing: FPS = 10.00, decoding = 1ms, frames = 600, invalid = 0], [INSTANCE0: FPS = 10.00, processed = 600]`

**Reply by awawa-dev** [reply]
Take a look in the pinned topic here. The power supply and possible common ground is one question. After connecting LED strip directly to the Rpi (and additionally without level shifter) you may experience problems like blinking/the LED strip does not work at full length/LED strip does not work at all. Simply, real time OS running on the Rpi is not good to emulate neopixel protocol which requires strictly narrow timings. The more leds (or using 4 byte sk6812 RGBW which provide very good quality, instead of 3 byte ws2812b RGB) the harder to provide stable communication. And if you are already running HyperHDR as a root, then you won't see any other errors in logs even if new LED strip refuses to work. The problem is beyond the HyperHDR/Rpi and we can do nothing to diagnose it further.

**Reply by FreubelNL** [reply]
Clear. So basicly the rpi isn't strong enough to deal with the data that has to be processed if one would go all out in number of leds and type of leds, hence the esp. I do notice the occasional flicker but i kinda thought it's because of how i connected everything with non ideal grabbers and hdmi. Well off to go buy more stuff. It's going to get crowded behind the tv lol.

**Reply by awawa-dev** [reply]
It's rather not about CPU power, Rpi is strong enough but other tasks coming from the OS can disrupt the neopixel emulation/communication. So external ESP specialized unit is highly recommended. Small blinking visible on dark scenes can also come from the grabber, but HyperHDR smoothing configuration has anti-flickering filter (with a timeout) that should be enabled and eliminate it. If the blinking is sudden and it's rather a burst of light, it's probably an electrical problem that esp with a level voltage shifter should solve.

**Reply by FreubelNL** [reply]
Ok. so far it's clear. I hope i'm not taking all of your time ! So i just purchased [these esp's](https://www.amazon.nl/gp/product/B074RGW2VQ/ref=ppx_yo_dt_b_asin_title_o00_s00?ie=UTF8&psc=1) but it gets a little complicated now. 
So it should now go like this:
receiver hdmi -> splitter -> usb grabber -> rpi. The rpi runs hyperhdr and that connects trough wifi to the esp32 which then controls the ledstrip by using pin 16 connected to the green wire on the ledstrip.
Then i tell hyperhdr in the controller type that it connects to wled and i tell wled that i'm using the sk6812 ledstrip.

**Reply by awawa-dev** [reply]
Still I can't see a level shifter here. I'm not a fan of wifi connection either, it's not a perfect driver for HyperHDR. There is a constant high-rate color stream which heavily depends on wifi quality signal because we don't use WLED just to turn on or off some fancy effect from time to time. Especially when you can have a cable connection to esp. As you can find it on my blog almost everything can be put into a plastic case: only HDMI splitter and probably USB grabber (depending on the model size) will be outside.

**Comment by FreubelNL** [comment]
So last night i made the plunge by reinstalling and connecting everything as far as i had the new items. The ESP32's aren't in yet but it looks like i don't need those after all for this project.
I ripped off the old unknown brand led strip and installed the new BTF ws2812b lighting strip.
Swopped out the USB grabber for the new HDMI splitter and the MS2109 knockoff usb grabber, and took it slow this time. Soldering all the connections and finish everything with heat shrink tubing. First run i was a little dissapointed but figured out that the hdmi splitter wasn't set correctly. After dealing with that it was a lot better, even better then the first set up so i was pretty happy. After getting used to the new situation i still had the feeling something was off. I rememberd the part about the LUT file and managed to get that into the RPI. Rebooted the RPI and wauw.... the ambilight is picture perfect. I can't think of anything that has made me happier with tech the last few years. So thank you very much for the help and patience you had with me !
It's good to know that the HDMI splitter and the ms2109 knock off work flawlessly.

**Comment by FreubelNL** [comment]
Hiya Awawa, i was just wondering, have you ever looked into the dig-uno ? (no worries, everything is still working as intended) 
It's a level-shifter and esp all in one. [Looks like everything i'm still missing](https://quinled.info/pre-assembled-quinled-dig-uno/)
Love to see your reaction.

**Reply by awawa-dev** [reply]
Hi
I know one user tested it and didn't work as intended with HyperSerialESP32: I don't have this board so I don't know if the problem was its modified version of HyperSerialESP32 firmware after tuning (the output pinout must be changed due to the board design) or incorrect custom HyperHDR configuration in this one particular example,  or it's a general problem with this board which have a conflict of resources used by HyperSerial.

**Reply by FreubelNL** [reply]
Would you like to try one ?

**Reply by awawa-dev** [reply]
I've got at least 10 ESP modules working fine at one of 1/2/4Mb speed with HyperSerial so thank you :) Probably someone else must test it again providing this time full feedback what she/he is changing in the HyperSerialESP32 firmware, debug output from the device and full HyperHDR's configuration&log to see what's going on. I can't see any reason why this ESP32 board with a level-shifter shell shouldn't work after customizing the HyperSerial firmware.

**Reply by FreubelNL** [reply]
Fair enough ! If you do find the time just give me a shout, i'd love to buy one for you to see if you can make it work with your awesome hyperhdr.

**Comment by ChadwickTheCrab** [comment]
> And in case of MS2109 the calibration may not be necessary. I made the calibration myself and it's shared to download in the HyperHDR release section. You must only follow the short manual inside the archive that will tell you, how to setup brightness/contrast/hue etc  in HyperHDR because it's important for this grabber (the factory settings are not neutral and distort the video).

Hey I love your software thank you for all your work! I'm having trouble trying to get the backlight threshold set at zero to go completely off during a full black scene. I'm assuming it's as if it's interpreting black as a dark grey. I can get them off by adjusting the hardware contrast but then it throws the colors way off. I have the MS2109 with your LUT installed. Brightness 1, contrast 128, saturation 128, and gamma 1.7 for each channel. Any tips? My TV isn't an HDR capable TV but my receiver is capable of HDR output to the capture, unsure if that is relevant.

**Reply by awawa-dev** [reply]
Hi
Thanks. First,  set `Backlight threshold` to 0 in the processing tab. You can also try to set grabber hardware brightness to -1.  If that doesn't help let HyperHDR v20 to download LUT for MS2130 **limited colorspace**: remember to restore Brightness 1, contrast 128, saturation 128 for the grabber because this LUT will reset them and MS2109 needs them.

**Reply by awawa-dev** [reply]
>  My TV isn't an HDR capable TV

LUT is also used for SDR if you use YUV format for the grabber. Alternatively you can switch to MJPEG to see if it resolves the issue (MJPEG doesnt use LUT for SDR so LUT is irrelevant for this case). No matter what you choose, you should disable HDR tone mapping because neither MS2109 or your TV wont enforce HDR signal (unless you force it using HDMI splitter for an unknown reason).

**Reply by ChadwickTheCrab** [reply]
> Hi
> Thanks. First,  set `Backlight threshold` to 0 in the processing tab. You can also try to set grabber hardware brightness to -1.  If that doesn't help let HyperHDR v20 to download LUT for MS2130 **limited colorspace**: remember to restore Brightness 1, contrast 128, saturation 128 for the grabber because this LUT will reset them and MS2109 needs them.
> 

Awesome I'll give that a shot if I don't like the results from tweaking your MS2109 lut. I think I could do a better job of adjusting gamma/contrast to better match the TV so when I adjust settings it's not exaggerating something already inaccurate. I was adjusting things based on what I would see in the preview on my phone (OLED), vs when I look at the interface off my laptop (IPS LED).

**Reply by awawa-dev** [reply]
The MS2109 captures SDR in good quality, so there is probably no need to modify the LUT or download and install a custom LUT. It may even be better to delete the custom LUT from your home folder if you have already installed it, so as not to complicate things.

But of course it is possible also to automatically calibrate the LUT also for SDR. The calibrator works fine not only for HDR. Then you will also find out how much the actual colors differ from what is to be displayed on the TV. For this grabber, the differences are usually negligible. And only in this case, if you have **your own LUT calibrated for SDR** (not downloaded, these are for HDR when tone mapping is enabled), you should enable tone mapping to activate the corrections.

**Reply by ChadwickTheCrab** [reply]
Now if I have adjusted my tv picture settings to get a subjectively better image to our tastes, shouldn't I try to adjust the grabber to match that instead of just being as true to possible to the input?

**Reply by awawa-dev** [reply]
ms2109 should capture SDR colors almost 1:1 with what the TV receives. Of course, the TV can use its own enhancers, e.g. my LG can use dynamic SDR to HDR tone mapping for a regular SDR signal, which completely changes the situation: we captured the signal almost perfectly, but the TV used its own enhancers. Or use a dynamic contract to highlight details. So what if we captured the signal if the TV processed it in its own way ;) Fortunately, such aggressive options are usually turned off by default. Since this is an SDR signal, you can also change the grabber settings, e.g. hardware brighness/contranst/saturation/hue correction, use a dedicated LUT or even experiment with LED calibration by that's my opinion: https://github.com/awawa-dev/HyperHDR/issues/853#issuecomment-2105664800 There are no strict rules here, set them to suit you best.
